posts/zzz_DO_NOT_EDIT_the__tensio.../the__tension_of__high-_performance__computing:__reproducibility_vs.__parallelization.qmd

```diff
@@ -0,0 +1,155 @@
+high speeds. As pharmaceutical research becomes increasingly data-driven, the need for powerful computational tools has grown, allowing for more accurate predictions, faster testing, and more efficient processes. However, with the growing complexity and scale of these simulations, ensuring reproducibility of results becomes a significant challenge.
+
+In this blog post, we will explore common reproducibility challenges in drug development and simulations, focusing on how the `{mirai}` package can be used as a backend solution to effectively manage parallelization.
+
+## The Problem: Reproducibility in Parallel Simulations
+
+Imagine a research team at the forefront of developing a new drug. They use sophisticated simulations to predict how the drug will perform across different patient cohorts. To manage the large computational workload, the team employs parallel processing, distributing the simulation tasks across multiple processors. This approach significantly speeds up the process, allowing them to handle vast datasets efficiently.
+
+However, the team soon encounters a problem. Each time they rerun the simulations, the results differ slightly, even though they use the same input parameters. This inconsistency raises a red flag: *their results are not reproducible.* In the pharmaceutical industry, where accuracy and reliability are paramount, this is a serious issue. Reproducibility is not just a scientific ideal; it's a regulatory requirement.
+
+Upon investigation, the team discovers that the variability in their results is due to the way tasks are parallelized across processors. The order in which operations are executed can differ slightly between runs, leading to small but significant variations in the outcomes. These differences are particularly problematic when they accumulate over thousands of iterations, making it difficult to ensure that the simulation results can be consistently reproduced by others.
+
+### Tracking Operations in Parallel Computing
+
+Let’s explore a simple scenario where parallelization creates confusion in tracking operations due to the asynchronous nature of task execution and logging.
+
+```{r, message=FALSE, eval=FALSE}
+library("mirai")
+library("dplyr", warn.conflicts = FALSE)
+
+# start parallel workers
+daemons(4)
+
+# load libraries on each worker and set up logging to a file
+everywhere({
+  library("dplyr")
+  library("tidyr")
+  library("mirai")
+  library("tidylog", warn.conflicts = FALSE)
+  
+  # Define function to log messages to the log file
+  log_to_file <- \(txt) cat(txt, file = log_file, sep = "\n", append = TRUE)
+  options("tidylog.display" = list(message, log_to_file))
+}, log_file = "log.txt")
+
+m <- mirai_map(letters[1:3], \(x) {
```

Comment:
Where does `mirai_map()` come from? I can't seem to find it when I load `library(mirai)` so the code chunk does not render for me.

Reply:
This function is for mirai >= 1.1.0
Can you check your version? 
From my side, I build the site using the docker file in the readme.